otaclient/app/ota_client.py

```diff
@@ -589,115 +615,189 @@ def status(self) -> wrapper.StatusResponseEcuV2:
-    Only one ongoing update/rollback is allowed, with the help of exclusive update_rollback lock.
-
-    If this ECU enables otaproxy and has child ECUs depend on it, control_flags is used to prevent
-    this ECU's reboot before the otaproxy dependency is resolved.
-
-    OTAClient itself is self-contained, its error handling is done internally, no exception will be raised
-    from calling API methods of OTAClient.
-    """
-
+class OTAServicer:
     def __init__(
         self,
         *,
-        ecu_info: ECUInfo,
-        executor: ThreadPoolExecutor,
         control_flags: OTAClientControlFlags,
+        ecu_info: ECUInfo,
+        executor: Optional[ThreadPoolExecutor] = None,
+        otaclient_version: str = __version__,
         proxy: Optional[str] = None,
     ) -> None:
-        # only one update/rollback is allowed at a time
-        self.update_rollback_lock = asyncio.Lock()
-        self.my_ecu_id = ecu_info.ecu_id
+        self.ecu_id = ecu_info.ecu_id
+        self.otaclient_version = otaclient_version
+        self.local_used_proxy_url = proxy
+        self.last_operation: Optional[wrapper.StatusOta] = None
+
+        # default boot startup failure if boot_controller/otaclient_core crashed without
+        # raising specific error
+        self._otaclient_startup_failed_status = wrapper.StatusResponseEcuV2(
+            ecu_id=ecu_info.ecu_id,
+            otaclient_version=otaclient_version,
+            ota_status=wrapper.StatusOta.FAILURE,
+            failure_type=wrapper.FailureType.UNRECOVERABLE,
+            failure_reason="unspecific error",
+        )
+        self._update_rollback_lock = asyncio.Lock()
         self._run_in_executor = partial(
             asyncio.get_running_loop().run_in_executor, executor
         )
 
-        # create otaclient instance and control flags
-        self._otaclient = OTAClient(
-            boot_control_cls=get_boot_controller(ecu_info.get_bootloader()),
-            create_standby_cls=get_standby_slot_creator(cfg.STANDBY_CREATION_MODE),
-            my_ecu_id=self.my_ecu_id,
-            control_flags=control_flags,
-            proxy=proxy,
-        )
+        #
+        # ------ compose otaclient ------
+        #
+        self._otaclient_inst: Optional[OTAClient] = None
 
-        # proxy used by local otaclient
-        # NOTE: it can be an upper otaproxy, local otaproxy, or no proxy
-        self.local_used_proxy_url = proxy
-        self.last_operation = None  # update/rollback/None
+        # select boot_controller and standby_slot implementations
+        _bootctrl_cls = get_boot_controller(ecu_info.get_bootloader())
+        _standby_slot_creator = get_standby_slot_creator(cfg.STANDBY_CREATION_MODE)
+
+        # boot controller starts up
+        try:
+            _bootctrl_inst = _bootctrl_cls()
+        except ota_errors.OTAError as e:
+            logger.error(
+                e.get_error_report(title=f"boot controller startup failed: {e!r}")
+            )
```

Comment:
boot controller is launched before otaclient core(OTAClient) starts, and we capture and handle the boot controller startup error here.

Error will be captured and parsed, error report(including full traceback) will be logged, failure reason will be available via status API. When DEBUG_MODE is on, full traceback will also be available via status API.